pgjdbc/src/main/java/org/postgresql/hostchooser/MultiHostChooser.java

```diff
@@ -56,32 +56,40 @@ public Iterator<CandidateHost> iterator() {
   }
 
   private Iterator<CandidateHost> candidateIterator() {
-    if (targetServerType != HostRequirement.preferSecondary) {
+    if (   targetServerType != HostRequirement.preferSecondary
+        && targetServerType != HostRequirement.preferPrimary   ) {
       return getCandidateHosts(targetServerType).iterator();
     }
 
+    HostRequirement preferredServerType =
+        targetServerType == HostRequirement.preferSecondary
+          ? HostRequirement.secondary
+          : HostRequirement.primary;
+
     // preferSecondary tries to find secondary hosts first
     // Note: sort does not work here since there are "unknown" hosts,
     // and that "unknown" might turn out to be master, so we should discard that
     // if other secondaries exist
-    List<CandidateHost> secondaries = getCandidateHosts(HostRequirement.secondary);
+    // Same logic as the above works for preferPrimary if we replace "secondary"
+    // with "primary" and vice versa
+    List<CandidateHost> preferred = getCandidateHosts(preferredServerType);
     List<CandidateHost> any = getCandidateHosts(HostRequirement.any);
 
-    if (secondaries.isEmpty()) {
+    if (preferred.isEmpty()) {
       return any.iterator();
     }
 
     if (any.isEmpty()) {
-      return secondaries.iterator();
+      return preferred.iterator();
     }
 
-    if (secondaries.get(secondaries.size() - 1).equals(any.get(0))) {
-      // When the last secondary's hostspec is the same as the first in "any" list, there's no need
-      // to attempt to connect it as "secondary"
+    if (preferred.get(preferred.size() - 1).equals(any.get(0))) {
+      // When the last preferred host's hostspec is the same as the first in "any" list, there's no need
+      // to attempt to connect it as "preferred"
       // Note: this is only an optimization
-      secondaries = rtrim(1, secondaries);
+      preferred = rtrim(1, preferred);
```

Comment:
`any.removeAll(preferred)` would avoid duplicate attempts even more effectively, and not require explanation

Reply:
I agree. I just tried to keep the changes minimal and keep the logic as close as possible to the original logic

Reply:
> `any.removeAll(preferred)` would avoid duplicate attempts even more effectively, and not require explanation

After further consideration I think that `any.removeAll(preferred)` won't work with `CandidateHost` class instances because it won't find a match in `any` for any instance from `preferred` and, even if implemented properly (i.e. against `HostSpec` instances instead of `CandidateHost`), then it will break the logic implemented here originally for the host selection.

> None of the checks are actually necessary:
> 
> ```
> List<CandidateHost> preferred = getCandidateHosts(preferredServerType);
> List<CandidateHost> any = getCandidateHosts(HostRequirement.any);
> 
> any.removeAll(preferred);
> return append(preferred, any).iterator();
> ```
> 
> May need a different function call if the lists aren't mutable. Or skip `removeAll` and don't worry about repeated attempts.

Those repeated attempts are actually important here since they are made for different `HostRequirement`s

Reply:
> won't work with `CandidateHost` class instances 

Ah yes. I missed the `.hostSpec` in the original. Removing all of them would need a bit more code.

> Those repeated attempts are actually important 

Then why is it removing one of them as an optimisation?

Reply:
> Then why is it removing one of them as an optimisation?

It is removing just this one for optimization because it would check this host two times in a row, first time for a narrower requirement/condition (e.g. `secondary`) and, if the first check fails, it checks it right away for the broader `any` requirement/condition. These two checks in a row don't make too much sense so it removes the check for the narrower condition and makes just one check for the broader `any` condition

Reply:
It's important that the `HostSpec`s' order in both parts of the returned `List<CandidateHost>` result (i.e. the first `preferred` part and the last `any` part) is the same as the order of host specs in the jdbc url. What's also important is that those `preferred` and `any` parts are shuffled independently, when `loadBalanceHosts=true`

Reply:
+1. The previous logic with trimming `secondaries` was invalid.